src/lib.rs

```diff
@@ -269,11 +270,14 @@ impl Header {
         let ptr = self as *const Header as *mut Header as *mut u8;
 
         unsafe {
-            if padding > 0 && self.len() == 0 {
+            if padding > 0 && self.cap() == 0 {
                 // The empty header isn't well-aligned, just make an aligned one up
                 NonNull::dangling().as_ptr()
             } else {
-                ptr.offset(header_size as isize) as *mut T
+                // This could technically result in overflow, but padding would
+                // have to be absurdly large for this to occur.
+                let offset = header_size + padding;
+                ptr.offset(isize::try_from(offset).expect("offset fits into isize")) as *mut T
```

Comment:
You shouldn't be doing checks like this here -- all of that checking should be done in `alloc_size`. (It has a TODO about caring about isize::MAX, but otherwise checks for any other kind of overflow).

Reply:
Also since this code was written `ptr.add` has been introduced to elide this annoying cast (but it still is contrained to isize because of backend limitations), I suggest using that instead.

Reply:
Done, thanks for the review!